Obey prefer_full_module setting when finding backreferences

QuLogic opened the pull request:
This fixes several cases of missing backreferences in Matplotlib's `mpl_toolkits` namespace, which need the full module name for referencing into Sphinx.

For example, [`Axes3D.plot`](https://matplotlib.org/stable/api/_as_gen/mpl_toolkits.mplot3d.axes3d.Axes3D.plot.html) only has one back reference:
![image](https://github.com/user-attachments/assets/8640693d-b73e-4d22-bff8-7ad060929ddc)

but with this change we have all these:
![image](https://github.com/user-attachments/assets/6712270f-7eca-4d75-9d49-b49ad6cdd303)

larsoner commented:
Would be great to have a test that would have caught this. @QuLogic WDYT about adding an Axes3D to one of the `sphinx_gallery/tests/tinybuild` examples? On `master` that wouldn't backref / minigallery properly but it would on this PR, right? If so then adding it there you could check in `sphinx_gallery/tests/test_full.py` somewhere that it works. (Maybe you'd need to add a custom minigallery to the build as well?)

QuLogic commented:
But for back references, it would have to be something in the documented module, so from `sphinx_gallery` for `tinybuild`, wouldn't it?

lucyleeow commented:
I think any module listed in `doc_module` would work, see: https://sphinx-gallery.github.io/dev/configuration.html#add-mini-galleries-for-api-documentation

QuLogic commented:
I added a `_dummy` module, and moved `DummyClass` there, along with a nested `NestedDummyClass`. And the added tests should use that and cover this and #1363.

QuLogic commented on sphinx_gallery/tests/test_full.py:
```diff
@@ -548,25 +548,49 @@ def test_embed_links_and_styles(sphinx_app):
     # instance
     dummy_class_inst = re.search(
-        r'sphinx_gallery.backreferences.html#sphinx[_,-]gallery[.,-]backreferences[.,-][D,d]ummy[C,c]lass" title="sphinx_gallery.backreferences.DummyClass" class="sphx-glr-backref-module-sphinx_gallery-backreferences sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">dc</span>',  # noqa: E501
+        r'sphinx_gallery._dummy.html#sphinx[_-]gallery[.-]_dummy[.-][Dd]ummy[Cc]lass" title="sphinx_gallery._dummy.DummyClass" class="sphx-glr-backref-module-sphinx_gallery-_dummy sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">dc</span>',  # noqa: E501
```
I also fixed the regex here, as #617 seems to be saying that `_` or `.` may be `-` instead, but these seem to be also allowing `,` which seems unlikely.

larsoner commented:
Yikes I thought that would be simpler, thanks for working through it @QuLogic !

lucyleeow commented:
Stupid question, could we have also tested this by adding mpl to `doc_modules` and adding a `minigallery` directive with `Axes3D.plot` ? (obv this seems more difficult than amending dummy)
